### PR TITLE
dev sealv4

### PR DESCRIPTION
**Pull request by devharsh**
- support for macOS
- SEALv4 upgrade with BGV scheme

### CODE REVIEW

File: Pyfhel/utils/cy_type_converters.pxi [hunk 1]
```diff
@@ -2,26 +2,28 @@ from Pyfhel.utils.Scheme_t import Scheme_t
...
             * (scheme_t) Enum (does nothing)
 
     Returns:
-        scheme_t: bfv or ckks.
+        scheme_t: bfv, bgv or ckks.
     """
     if type(scheme) is type: scheme = scheme.__class__.__name__
     if type(scheme) is str:
         if scheme.lower() in ('int', 'integer', 'bfv'):     scheme = "bfv"
+        elif scheme.lower() in ('int', 'integer', 'bgv'):   scheme = "bgv"
```
Comment (ibarrond) [inline Pyfhel/utils/cy_type_converters.pxi:26]: Sadly we cannot have two schemes named 'int' or 'integer'. To keep backwards compatibility, I'd say we only allow BGV scheme to be referenced as 'bgv', leaving all integer-like references for BFV.

Change this line to `        elif scheme.lower() == 'bgv':   scheme = "bgv"`, and update comments related to it.

File: Pyfhel/utils/cy_type_converters.pxi [hunk 2]
```diff
@@ -2,26 +2,28 @@ from Pyfhel.utils.Scheme_t import Scheme_t
...
 cpdef to_Scheme_t(object scheme):
-    """Turns `scheme` into an scheme_t.{bfv, ckks} enum.
+    """Turns `scheme` into an scheme_t.{bfv, bgv, ckks} enum.
     
     Arguments:
         scheme (str, type, int, scheme_t): One of the following:
 
-            * (str): ('int', 'integer', 'bfv') for bfv scheme, 
+            * (str): ('int', 'integer', 'bfv') for bfv scheme,
+                     ('int', 'integer', 'bgv') for bgv scheme,
```
Comment (ibarrond) [inline Pyfhel/utils/cy_type_converters.pxi]: Only 'bgv'

File: Pyfhel/Pyfhel.pyx [hunk 3]
```diff
@@ -239,6 +241,15 @@ cdef class Pyfhel:
...
             assert (sec>0 or q>0), "BFV scheme requires `sec` or `q` to be set."
             self._scale = 1
+        elif s==Scheme_t.bgv:
+            assert (t_bits>0 or t>0), "BGV scheme requires `t_bits` or `t` to be set"
+            if not qi.empty():  # Compress all moduli into one
+                q = np.prod(np.array(qi))
+                self._qi = qi
+            else:
+                self._qi = {}
+            assert (sec>0 or q>0), "BFV scheme requires `sec` or `q` to be set."
```
Comment (ibarrond) [inline Pyfhel/Pyfhel.pyx:251]: BGV instead of BFV

File: Pyfhel/PyPtxt.pyx [hunk 4]
```diff
@@ -248,6 +249,8 @@ cdef class PyPtxt:
         """returns the integer in the first slot of the plaintext"""
         if (self._scheme != scheme_t.bfv):
```
Comment (ibarrond) [inline Pyfhel/PyPtxt.pyx:250]: scheme_t.bgv would raise this exception. change to have both conditions in one single `if` clause (`if a and b`)

File: Pyfhel/PyPtxt.pyx [hunk 5]
```diff
@@ -228,6 +228,7 @@ cdef class PyPtxt:
             content: (:obj:`bytes`) Python bytes object containing the PyPtxt.
             scheme: (:obj: `str`) String or type describing the scheme:
               * ('int', 'integer', int, 1, scheme_t.bfv) -> integer scheme.
+              * ('int', 'integer', int, 1, scheme_t.bgv) -> integer scheme.
```
Comment (ibarrond) [inline Pyfhel/PyPtxt.pyx:231]: Change to ` (3, scheme_t.bgv) -> integer scheme.`. We will leave BFV as default integer scheme.

File: Pyfhel/PyCtxt.pyx [hunk 6]
```diff
@@ -723,6 +729,10 @@ cdef class PyCtxt:
             other = other.astype(np.int64)
             t = self._pyfhel.t
             inverse = np.array([pow(int(other[i]), t-2, t)for i in range(other.shape[0])])
+        elif self.scheme == Scheme_t.bgv:
```
Comment (ibarrond) [inline Pyfhel/PyCtxt.pyx:732]: Merge with previous `if` clause, since they do the same thing! (`if a or b`).

File: Pyfhel/PyCtxt.pyx [hunk 7]
```diff
@@ -621,6 +622,9 @@ cdef class PyCtxt:
         if self.scheme==Scheme_t.bfv:
             scheme_dep_info = 'noiseBudget=' + str(self.noiseBudget) \
                                             if self.noiseBudget!=-1 else "?"
+        elif self.scheme==Scheme_t.bgv:
```
Comment (ibarrond) [inline Pyfhel/PyCtxt.pyx:625]: Merge with previous `if` clause, since they do the same thing! (`if a or b`).

File: Pyfhel/PyCtxt.pyx [hunk 8]
```diff
@@ -246,6 +246,7 @@ cdef class PyCtxt:
             scheme (str, type, int, Scheme_t): One of the following:
 
                 * ('int', 'INTEGER', int, 1, Scheme_t.bfv) -> integer scheme.
+                * ('int', 'INTEGER', int, 1, Scheme_t.bgv) -> integer scheme.
```
Comment (ibarrond) [inline Pyfhel/PyCtxt.pyx:249]: `(3, Scheme_t.bgv) -> integer scheme.`

File: Pyfhel/utils/Scheme_t.py [hunk 9]
```diff
@@ -9,4 +9,8 @@ class Scheme_t(Enum):
     """Integer encoding, used with encryptInt/decryptInt and encodeInt/decodeInt."""
 
     ckks=0x2
-    """Fractional encoding, used with encryptFrac/decryptFrac and encodeFrac/decodeFrac."""
\ No newline at end of file
+    """Fractional encoding, used with encryptFrac/decryptFrac and encodeFrac/decodeFrac."""
+
+    bgv=0x3
+    """Integer encoding, used with encryptInt/decryptInt and encodeInt/decodeInt."""
```
Comment (ibarrond) [inline Pyfhel/utils/Scheme_t.py:15]: used with encryptBGV/...

File: examples/Demo_BGV.py [hunk 10]
```diff
@@ -1,163 +1,81 @@
 """
-Integer FHE with BGV scheme
-========================================
+HelloWorld
```
Comment (ibarrond) [inline examples/Demo_BGV.py:2]: Merge the two BGV demos into one, and name it `Demo_7_Integer_BGV.py` .

File: examples/Demo_BGV2.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
+"""
```
Comment (ibarrond) [inline examples/Demo_BGV2.py:1]: Merge with Demo_BGV

**Review by ibarrond — CHANGES_REQUESTED** [review]

**Review by ibarrond — APPROVED** [review]

**Comment by ibarrond** [comment]
A few things to change before merging:
1. Pyfhel==3.3.0 introduced a bunch of changes that conflict with this PR, and thus they need to be resolved.
2. The pull request still shows thousands of lines changed due to an issue with the newline. While this can be filtered out in Github, git is not that smart by itself. Remove all these unnecessary changes and rebase the entire pull request so that only the important changes are committed.
3. MacOS support is based on setting global environment variables, but this can be compiler-dependent and not controlled by Python. Review the approach to add compiler and linker directives only in the `pyproject.toml` file and avoid using environment variables.
4. Update Demo_7_BGV to Demo_9_BGV.
5. Add coverage tests to the new pieces of code.

**Comment by davidhassel** [comment]
Any updates on this? We are trying test BGV, but have some issues with the installation (compilation errors in Pyfhel/Pyfhel.cpp when building 'Pyfhel.Pyfhel' extension) (e.g., `Pyfhel/Pyfhel.cpp:26069:64: error: cannot convert ‘std::vector<long long int>’ to ‘std::vector<long int>&’` when running `pip install .` However, not sure what stage this pull request is at?

@devharsh @ibarrond

**Comment by ibarrond** [comment]
@davidhassel BGV is now implemented in the `dev` branch, building and testing correctly for Windows and Ubuntu. You can install it from there.

**Comment by ibarrond** [comment]
@devharsh To avoid the full rewrite due to the newline issue you had, I mocked your commits in the `dev` branch and kept updating it there. Everything related to BGV now works in Windows/Linux, and I'm stuck with the installation in MacOs (testing it automatically in a fresh MacOs environment via Github Actions). If you have the time, let's move the discussion to #131 to solve the last missing issues.

Closing this branch!

**Comment by devharsh** [comment]
Sure, always happy to help!

Best,
Devharsh.

On Dec 2, 2022, at 3:32 PM, Alberto Ibarrondo ***@***.***> wrote:

﻿

@devharsh<https://github.com/devharsh> To avoid the full rewrite due to the newline issue you had, I mocked your commits in the dev branch and kept updating it there. Everything related to BGV now works in Windows/Linux, and I'm stuck with the installation in MacOs (testing it automatically in a fresh MacOs environment via Github Actions). If you have the time, let's move the discussion to #131<https://github.com/ibarrond/Pyfhel/issues/131> to solve the last missing issues.

Closing this branch!

**Pull request closed without merging**